File: rallye-webapp/src/app/configuration/models/challenge-configuration.ts
import { HalLinks } from '../../models/hal-links';
import { PerformanceScoring } from './performance-scoring';
import { QuestionDefinition } from './question-definition';
import { QuestionScoring } from './question-scoring';
import { ChallengeGroup } from './challenge-group';

export interface QuestionScorings {
  [key: string]: QuestionScoring;
}

export interface PerformanceScorings {
  [key: string]: PerformanceScoring;
}

export interface QuestionDefinitions {
  [key: string]: QuestionDefinition;
}

export interface ChallengeConfiguration {
  id?: string;
  version?: number;
  /** Nom historique du numéro d'épreuve dans le contrat JSON du back. */
  challenge: number;
  name: string;
  questionScorings: QuestionScorings;
  performanceScorings: PerformanceScorings;
  questionDefinitions: QuestionDefinitions;
  group?: ChallengeGroup | null;
  _links?: HalLinks;
}

File: rallye-core-service/src/main/java/fr/vandriessche/rallyeschema/coreservice/entities/ChallengeConfiguration.java
package fr.vandriessche.rallyeschema.coreservice.entities;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Version;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@Document
public class ChallengeConfiguration {
	public ChallengeConfiguration(Integer challenge) {
		this.challenge = challenge;
	}

	@Id
	private String id;

	@Version
	private Long version;

	@Indexed(unique = true)
	private Integer challenge;

	private String name;

	@DBRef
	private ChallengeGroup group;

	@DBRef
	private List<FormRecognitionConfiguration> formRecognitionConfigurations = new ArrayList<>();

	private LinkedHashMap<String, QuestionScoring> questionScorings = new LinkedHashMap<>();
	private LinkedHashMap<String, PerformanceScoring> performanceScorings = new LinkedHashMap<>();

	private LinkedHashMap<String, QuestionDefinition> questionDefinitions = new LinkedHashMap<>();
}
